### PR TITLE
Incorporate heapster-saw into saw-script

### PR DESCRIPTION
**Pull request by glguy**

### CODE REVIEW

File: .github/ci-heapster.sh [hunk 1]
```diff
@@ -46,7 +46,7 @@ setup_dist_bins() {
   fi
   extract_exe "saw" "dist/bin"
   export PATH=$PWD/dist/bin:$PATH
-  echo "$PWD/dist/bin" >> "$GITHUB_PATH"
+  echo "$PWD/dist/bin" >> $GITHUB_PATH
```
Comment (atomb) [inline .github/ci-heapster.sh:49]: Is this fixing the PATH issue we've previously had, or undoing the fix we put in place? I wonder if @ldettwy knows which we should be using.

File: .github/ci-heapster.sh [hunk 2]
```diff
@@ -105,6 +105,15 @@ install_yices() {
...
 
+install_yasm() {
+  is_exe "$BIN" "yasm" && return
+  if [[ "$RUNNER_OS" = "Linux" ]]; then
+    sudo apt-get update -q && sudo apt-get install -y yasm
+  else
+    brew install yasm
+  fi
+}
+
```
Comment (atomb) [inline .github/ci-heapster.sh:116]: This makes me think the above may be undoing a fix, since we no longer need YASM. I also notice that this PR _moves_ this file, which would break the existing CI. Maybe it makes more sense to modify the workflow file to use `ci.sh` and get rid of the changes that created `ci-heapster.sh`?

File: .github/workflows/heapster.yml [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  pull_request:
+    branches: [wip-heapster]
+
+jobs:
+  build:
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-latest, macos-latest]
+        ghc: ["8.6.5", "8.8.4", "8.10.2"]
```
Comment (atomb) [inline .github/workflows/heapster.yml:14]: Could this use 8.10.4 and avoid the need of a config file for 8.10.2?

File: .github/workflows/heapster.yml [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            dist-newstyle
+          key: cabal-${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('**/cabal.GHC-*') }}-${{ github.sha }}
+          restore-keys: |
+            cabal-${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('**/cabal.GHC-*') }}
+
+      - shell: bash
+        run: .github/ci-heapster.sh build_abc
+
+      - shell: bash
+        run: .github/ci-heapster.sh build
```
Comment (atomb) [inline .github/workflows/heapster.yml:68]: Based on my previous comments, this (and the rest of the similar things in this file) could use `ci.sh`.

File: cabal.GHC-8.10.2.config [hunk 5]
```diff
@@ -0,0 +1,294 @@
+constraints: any.Cabal ==3.2.0.0,
```
Comment (atomb) [inline cabal.GHC-8.10.2.config:1]: As mentioned previously, this file may not be needed.

File: cabal.GHC-8.8.4.config [hunk 6]
```diff
@@ -93,6 +93,7 @@ constraints: any.Cabal ==3.0.1.0,
              any.cryptohash-md5 ==0.11.100.1,
              any.cryptohash-sha1 ==0.11.100.1,
              cryptol +relocatable -static,
+             cryptol-saw-core +build-css,
```
Comment (atomb) [inline cabal.GHC-8.8.4.config:96]: Is this a necessary change?

File: cabal.GHC-8.8.4.config [hunk 7]
```diff
@@ -147,7 +148,7 @@ constraints: any.Cabal ==3.0.1.0,
              any.gitrev ==1.3.1,
              any.happy ==1.20.0,
              any.hashable ==1.3.1.0,
-             hashable +integer-gmp,
+             hashable -examples +integer-gmp +sse2 -sse41,
```
Comment (atomb) [inline cabal.GHC-8.8.4.config:151]: What's the motivation for this change? Is it consistent in the other freeze files?

File: cabal.project [hunk 8]
```diff
@@ -37,3 +38,9 @@ packages:
     deps/argo/argo
     deps/argo/tasty-script-exitcode
     deps/cryptol/cryptol-remote-api
+    deps/language-rust
+
+source-repository-package
```
Comment (atomb) [inline cabal.project:43]: I've been wondering whether we might want to switch to this mode for all of our git-located dependencies, but it's a bit weird to have a mix of this and submodules. Maybe this could be made into a submodule for now, for consistency, and then we could experiment with putting all of our git dependencies in `cabal.project` in a later PR.

File: heapster-saw/LICENSE [hunk 9]
```diff
@@ -0,0 +1,30 @@
+Copyright (c) 2013-2019 Galois Inc.
```
Comment (atomb) [inline heapster-saw/LICENSE:1]: Should we update this to 2021?

File: src/SAWScript/Proof.hs [hunk 10]
```diff
@@ -84,6 +84,7 @@ module SAWScript.Proof
 
   , CEX
   , ProofResult(..)
+
```
Comment (atomb) [inline src/SAWScript/Proof.hs:87]: I like to avoid unnecessary diffs.

**Review by atomb — CHANGES_REQUESTED** [review]
I think this mostly looks great, and I'm very happy that it can be incorporated without any changes to shared infrastructure. I made a few small comments of things I think we should change before merging to avoid build-related awkwardness with the rest of SAW, but they should be pretty easy. (I should note that I haven't carefully reviewed the files that are purely additions, but I think it makes sense to get them merged regardless of their state, given that they just add experimental commands and don't otherwise change functionality.)

File: .github/ci-heapster.sh (hunk 1 above)
Comment (glguy) [reply .github/ci-heapster.sh]: This appeared to be reverting a fix and I've added the quotes back.

File: .github/ci-heapster.sh (hunk 2 above)
Comment (glguy) [reply .github/ci-heapster.sh:116]: yes, I think you're right. I wasn't focused on how CI had been modified in heapster

File: cabal.project (hunk 8 above)
Comment (glguy) [reply cabal.project:43]: I'll ask about getting a hobbits release.

File: cabal.GHC-8.8.4.config (hunk 6 above)
Comment (eddywestbrook) [reply cabal.GHC-8.8.4.config:96]: I would guess that this is an artifact of multiple merges...? I have no idea what it is

File: cabal.GHC-8.8.4.config (hunk 7 above)
Comment (eddywestbrook) [reply cabal.GHC-8.8.4.config:151]: Also no idea what this one is about either. As the previous one, I would guess this is an artifact of past merges...?

File: cabal.project (hunk 8 above)
Comment (eddywestbrook) [reply cabal.project:43]: For `language-rust`, I think it might be being used by another SAW submodule...? I don't know. I think we are just using the version on hackage, so maybe it doesn't need to be in `deps`, but I am not completely sure.

As for Hobbits, I could imagine making it a submodule. I was originally hoping that we would update it and be done, but it does seem like we have been doing some amount of development on it. More generally, I would like (in the next, maybe, 12 months) to write another Hobbits paper, so I would like to make it release-able at some point, but that is not necessarily incompatible with making it a submodule, just something to consider.

**Comment by robdockins** [comment]
The history here is pretty messy.  There are sequences of patches back-to-back that just bump submodules.  I think it would be nice to squash those together in a rebase.

File: .gitmodules [hunk 11]
```diff
@@ -40,6 +40,12 @@
 [submodule "deps/what4"]
 	path = deps/what4
 	url = https://github.com/GaloisInc/what4.git
+[submodule "deps/heapster-saw"]
```
Comment (robdockins) [inline .gitmodules:43]: Do we still need this submodule?

**Comment by glguy** [comment]
> The history here is pretty messy. There are sequences of patches back-to-back that just bump submodules. I think it would be nice to squash those together in a rebase.

I think when we're happy with where we are we can squash the whole series of commits reconciling changes for the merge. Would you prefer a finer granularity than that?

File: cabal.GHC-8.8.4.config (hunk 7 above)
Comment (glguy) [reply cabal.GHC-8.8.4.config:151]: If we don't know why these flags were turned on, then we can probably just turn them back off

**Comment by robdockins** [comment]
> > The history here is pretty messy. There are sequences of patches back-to-back that just bump submodules. I think it would be nice to squash those together in a rebase.
> 
> I think when we're happy with where we are we can squash the whole series of commits reconciling changes for the merge. Would you prefer a finer granularity than that?

I was thinking about the main body of the branch, too; there's a lot of bumping various submodules, some of which don't really need to exist anymore.  My main concern is basically about being able to sensibly do bisection later.  I know @brianhuffman has some pretty strong opinions about this stuff too.

Personally, I'd be happy if the whole PR got squashed into a couple of big commits (one to add the heapster-saw subdirectory, another to tie it into SAW).  I suppose it depends on how valuable we view the overall history here.

**Comment by glguy** [comment]
> Personally, I'd be happy if the whole PR got squashed into a couple of big commits (one to add the heapster-saw subdirectory, another to tie it into SAW). I suppose it depends on how valuable we view the overall history here.

I think most of the interesting history is going to be trapped in the heapster-saw repository that's the main thing being merged in. The saw-script changes are likely less interesting to track. @eddywestbrook might have more to say on this.

**Comment by eddywestbrook** [comment]
I guess I don't have a strong opinion about how much history to capture, especially since most of the development is in the heapster-saw repo as @glguy suggested. There have certainly been breaking changes in the API of heapster-saw that have required matching changes in saw-script, but I don't think it is crucial to track all of that, especially as a lot of that had to do with matching changes to the Prelude in the saw-core repo, which no longer exists.

To sum up, any squashing on saw-script changes seems fine to me.

**Comment by ldettwy** [comment]
Notes from this morning:
- Merge strategy will be squash & archive, so the history will be kept in the repository (perhaps renamed?) and mentioned in the merge request description and merge commit message. I'll take care of the actual git merging for this.
- Thoughts on branch name, either for this or for things like this that we'd like to keep around but not be directly attached to `master`'s git history? I'd suggest keeping it simple and calling it `heapster`.
- Actions integration is still a WIP and I'll spend some time working through it today.

**Comment by eddywestbrook** [comment]
Well, all the CI passes. Can we finally merge this sucker?

**Review by ldettwy — APPROVED** [review]
Let's do this
